File: 28.implement-str-str.ts
/*
 * @lc app=leetcode id=28 lang=typescript
 *
 * [28] Implement strStr()
 */

// @lc code=start
function strStr(haystack: string, needle: string): number {
    function prefixFunction(s: string): number[] {
        // the process is to match the template string against itself.
        // a.k.a. the haystack is the needle.

        // pi array stores the length of the longest proper prefix that is also the suffix of s[:q]
        let pi: number[] = [];
        pi[0] = 0;
        // k has multiple meaning:
        // 1. the length of the longest *proper* prefix that is the suffix of s[:q], which is also
        // 2. the number of matched characters, because we are using s to match against itself.
        // 3. the index of current testing char of the template string
        let k = 0; 
        for (let q = 1; q < s.length; q += 1) {
            while (k > 0 && s.charAt(k) !== s.charAt(q)) { // using the 3rd meaning.
                k = pi[k-1] // the left k keeps using the 1st/2nd, whereas the right k uses the 3rd. 
            }

            if (s.charAt(k) === s.charAt(q)) { // using the 3rd meaning.
                k += 1; // the 1st meaning
            }
            pi[q] = k; // the 2nd meaning
        }
        return pi;
    }

    let pi: number[] = prefixFunction(needle); // index i stores the longest proper prefix that is also the suffix of needle[:i]
    let numberOfMatch = 0; // which is also the index of current testing char of the needle
    for (let i = 0; i < haystack.length; i += 1) {
        while (numberOfMatch > 0 && needle.charAt(numberOfMatch) !== haystack.charAt(i)) {
            numberOfMatch = pi[numberOfMatch-1];
        }
        if (needle.charAt(numberOfMatch) === haystack.charAt(i)) {
            numberOfMatch += 1;
        }

        if (numberOfMatch === needle.length) return i + 1 - needle.length;
    }

    return -1;
};
// @lc code=end

File: cpp/28.implement-str-str.cpp
#include <vector>
#include <string>

using namespace std;

class Solution {
public:
    int strStr(string haystack, string needle) {
        vector<int> prefixFunction(needle.size());
        getPrefixFuncion(prefixFunction, needle);

        int q = 0;

        for (int i = 0; i < haystack.size(); i++) {
            while (q > 0 && haystack[i] != needle[q])
                q = prefixFunction[q-1];

            if (haystack[i] == needle[q])
                q++;
            
            if (q == needle.size())
                return i - q + 1;
        }

        return -1;
    }

private:
    void getPrefixFuncion(vector<int> &prefixFunction, const string &s) {
        prefixFunction[0] = 0;
        int k = 0;

        for (int i = 1; i < s.size(); i++) {
            while (k > 0 && (s[i] != s[k]))
                k = prefixFunction[k-1];

            if (s[i] == s[k])
                k++;

            prefixFunction[i] == k;
        }
    }
};